Pass additional information to downward API (k8s version, kc version, k8s g/v)

neil-hickey opened the pull request:
Signed-off-by: Neil Hickey <nhickey@vmware.com>

<!--  Thanks for sending a pull request!  Here are some tips for you:

If this is your first time, please read our contributor guidelines: https://github.com/vmware-tanzu/carvel-kapp-controller/blob/develop/CONTRIBUTING.md and developer guide https://github.com/vmware-tanzu/carvel-kapp-controller/blob/develop/docs/dev.md
-->

#### What this PR does / why we need it:

#### Which issue(s) this PR fixes:
<!--
If no issue exists for this change, please create an issue and link it here.
-->
Fixes #781

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. 

-->
```release-note
Templating stages (ytt, helm, cue) now have the kubernetes version, available group/version APIs and kapp-controller version passed to them via data values, and enabled capabilites in helm.
```

#### Additional Notes for your reviewer:

##### Review Checklist:

- [x] Follows the [developer guidelines](https://carvel.dev/shared/docs/latest/development_guidelines/)
- [x] Relevant tests are added or updated
- [x] Code is at least as readable and maintainable as it was before this
  change
- [ ] Relevant docs in this repo added or updated
- [ ] Relevant carvel.dev docs added or updated in a separate PR and there's
  a link to that PR

#### Additional documentation e.g., Proposal, usage docs, etc.:

```docs

```

neil-hickey committed:
WIP: Retrieve version information prior to fet,temp,deploy steps

Signed-off-by: Neil Hickey <nhickey@vmware.com>

neil-hickey commented on pkg/app/app_reconcile_test.go:
```diff
@@ -223,3 +225,16 @@ func Test_TemplateError_DisplayedInStatus_UsefulErrorMessageProperty(t *testing.
 		fmt.Sprintf("Status is not same:\nExpected:\n%#v\nGot:\n%#v\n", expectedStatus, crdApp.app.Status()))
 }
+
+type FakeComponentInfo struct {
```
For now this is kinda duplicated across the tests.... Probably a better centralized place for it...

neil-hickey commented on pkg/app/app.go:
```diff
@@ -4,6 +4,7 @@
 package app
 
 import (
+	"github.com/vmware-tanzu/carvel-kapp-controller/pkg/componentInfo"
```
this behaviour of GoLand is annoying.. ignore the import formatting, I will go fix it later before proper review

neil-hickey committed:
Fix linter / import statements

Signed-off-by: Neil Hickey <nhickey@vmware.com>

joaopapereira commented on cmd/controller/run.go:
```diff
@@ -17,9 +17,10 @@ import (
 	pkgclient "github.com/vmware-tanzu/carvel-kapp-controller/pkg/apiserver/client/clientset/versioned"
 	"github.com/vmware-tanzu/carvel-kapp-controller/pkg/app"
 	kcclient "github.com/vmware-tanzu/carvel-kapp-controller/pkg/client/clientset/versioned"
+	"github.com/vmware-tanzu/carvel-kapp-controller/pkg/componentInfo"
```
```suggestion
	"github.com/vmware-tanzu/carvel-kapp-controller/pkg/componentinfo"
```

Changing the `I` to `i` since we try to keep all the package names lowercase. This comment will apply to all imports and also the folder names

joaopapereira commented on pkg/componentInfo/component_info.go:
```diff
@@ -0,0 +1,88 @@
+)
+
+// Info provides information about components of the system
+type Info interface {
```
I would move this interface into `pkg/app/app.ComponentInfo`.
Why: Interfaces generally should be nearer to the place where they are used. This allows you to separate implementation from the needs of a particular package. In this case, we will use all the functions on the interface but in other parts that might not happen.

joaopapereira replied:
As a rule of thumb that I use, I only store interfaces in the same package that declares a struct that implements them when:
1. This would be the only package consuming this interface/struct (very rare in occurrence)
2. This package is part of a library that is exported for others to use and implement their own structs. Normal case an example of that is [this library](https://github.com/vmware-tanzu/carvel-imgpkg/blob/develop/vendor/github.com/cppforlife/go-cli-ui/ui/interfaces.go#L7) 
3. There is some sort of dependency that can be circular an example of that is [this interface](https://github.com/vmware-tanzu/carvel-imgpkg/blob/develop/pkg/imgpkg/internal/util/level_logger.go#L10). In this example the function [NewIndentedLevelLogger](https://github.com/vmware-tanzu/carvel-imgpkg/blob/develop/pkg/imgpkg/internal/util/level_logger.go#L33) constructs a new logger based on another logger

joaopapereira commented on pkg/deploy/kapp.go:
```diff
@@ -16,7 +15,9 @@ import (
 	"github.com/vmware-tanzu/carvel-kapp-controller/pkg/exec"
+	"github.com/vmware-tanzu/carvel-kapp-controller/pkg/kubeconfig"
 	"github.com/vmware-tanzu/carvel-kapp-controller/pkg/memdir"
+	"gopkg.in/yaml.v2"
```
What are we using this package for? (sorry couldn't find it.) Also doesn't Kubernetes use their own yaml library?

joaopapereira commented on pkg/componentInfo/component_info.go:
```diff
@@ -0,0 +1,88 @@
+}
+
+// NewComponentInfo returns a ComponentInfo
+func NewComponentInfo(coreClient kubernetes.Interface, clusterAccess *kubeconfig.Kubeconfig, kappControllerVersion string) *ComponentInfo {
```
The `clusterAccess` parameter could also be an interface.

joe-kimmel-vmw commented on cmd/controller/run.go:
```diff
@@ -170,20 +171,22 @@ func Run(opts Options, runLog logr.Logger) error {
-	deployFactory := deploy.NewFactory(coreClient, kcConfig, sidecarCmdExec, runLog)
+	// initialize cluster access once - it contains a service account token cache which should be only setup once.
+	kubeconf := kubeconfig.NewKubeconfig(coreClient, runLog)
+	compInfo := componentInfo.NewComponentInfo(coreClient, kubeconf, Version)
```
lol ok i know "real go programmers don't use whole words" or something,  but this definitely feels like i'm suddenly in the HR portal checking on my compensation info.   anyways it's probably unambiguous given the context.

joe-kimmel-vmw commented on pkg/componentInfo/component_info.go:
```diff
@@ -0,0 +1,88 @@
+func (ci *ComponentInfo) KubernetesVersion(serviceAccountName string, specCluster *v1alpha1.AppCluster, objMeta *metav1.ObjectMeta) (semver.Version, error) {
+	switch {
+	case len(serviceAccountName) > 0:
+		v, err := ci.coreClient.Discovery().ServerVersion()
```
here on this line,  we're always going to call the same `coreClient.Discovery` , right?   I wonder if we could cache this call the first time we make it so like :

`ci.GetK8sVersion()`

and then 
```
func ci.GetK8sVer() (semver.Version, error){
  if ci.cachedVer != "" {
     return ci.cachedVer, nil
 }
  // the logic that you currently have, but then store the result into ci.cachedVer, which i just made up.
```


point being i _think_ it's a legit assumption that if we're calling th version of the cluster we're hosted on, it can't change without us getting restarted, so we can cache the version of our own cluster and thus minimize API calls.

joe-kimmel-vmw commented on pkg/componentInfo/component_info.go:
```diff
@@ -0,0 +1,88 @@
+			return semver.Version{}, err
+		}
+
+		v, err := clientset.Discovery().ServerVersion()
```
here we could contemplate whether we could do a cache based on the clientset, but i think we would have no way of subscribing to like, "tell me when the cluster i'm managing gets upgraded" , yeah? so might be a lot harder.

one ~heuristic is to keep a cache that's valid for e.g. ~1 minute or some short duration at a time, so you basically say, "ok but if there's a ton of these all at once (e.g. when the controller restarts on an existing cluster and has to re-reconcile everything) then i'll limit my API requests"

neil-hickey replied:
interesting! That means we would need to save this cache to the disk - for the usecase of restarts.

joe-kimmel-vmw commented on pkg/componentInfo/component_info.go:
```diff
@@ -0,0 +1,88 @@
+
+// KubernetesAPIs returns the available kubernetes Group/Version resources
+func (ci *ComponentInfo) KubernetesAPIs() ([]string, error) {
+	groups, err := ci.coreClient.Discovery().ServerGroups()
```
similar question/comment as above - i think this is hard for us to generally cache, but is it worth caching for short time durations (on order 1 minute) since generally this kind of thing doesn't change very often.

joe-kimmel-vmw commented on pkg/deploy/factory.go:
```diff
@@ -34,69 +28,40 @@ type KappConfiguration interface {
+	clusterOpts *v1alpha1.AppCluster, cancelCh chan struct{}, location kubeconfig.AccessLocation) (*Kapp, error) {
 
-	processedGenericOpts, err := f.processOpts(saName, clusterOpts, genericOpts)
+	clusterAccess, err := f.kubeconfig.ClusterAccess(saName, clusterOpts, location)
```
congratulations on refactoring away that `processedGenericOpts` biz.

joe-kimmel-vmw commented on pkg/componentInfo/component_info.go:
```diff
@@ -0,0 +1,88 @@
+// If AppSpec points to external cluster, we use that k8s version instead
+func (ci *ComponentInfo) KubernetesVersion(serviceAccountName string, specCluster *v1alpha1.AppCluster, objMeta *metav1.ObjectMeta) (semver.Version, error) {
+	switch {
+	case len(serviceAccountName) > 0:
```
it's slightly a bummer that this ~unintuitive check is still duplicated here and in the ClusterAccess code.  I'm guessing you already wrestled with this knot and this was the best solution you saw.  maybe worth thinking if there's a simplification that puts this check into exactly one place.

neil-hickey replied:
yup, I couldn't really find one sadly. Even if we refactored it out, we still *need* the branching in both places (though maybe it could be a shared function? I dunno if that helps)

joe-kimmel-vmw commented on pkg/componentInfo/component_info.go:
```diff
@@ -0,0 +1,88 @@
+		}
+		return semver.ParseTolerant(v.String())
+	case specCluster != nil:
+		accessInfo, err := ci.clusterAccess.ClusterAccess(serviceAccountName, specCluster, kubeconfig.AccessLocation{Name: objMeta.Name, Namespace: objMeta.Namespace})
```
hmmm especially considering that in the second case here we _do_ delegate to ClusterAccess.   ...  is there a way to delegate the whole thing to cluster access so that there's only one file that checks the saname length, even if that one file has two codepaths that check it?    i guess that would look like moving the bulk of this function into the clusterAccess ...   how would that feel?

joe-kimmel-vmw commented on pkg/packageinstall/packageinstall.go:
```diff
@@ -248,12 +236,14 @@ func (pi *PackageInstallCR) kcVersionConstraintsSatisfied(pkg *datapkgingv1alpha
 	}
 
-	v, err := dropVersionPreReleaseAndBuild(pi.controllerVersion)
+	v, err := pi.compInfo.KappControllerVersion()
```
this is cleaner, nice

joe-kimmel-vmw commented on pkg/template/downward_api_values.go:
```diff
@@ -9,33 +9,69 @@ import (
+			}
+			keyValueContent, err = a.extractFieldPathAsKeyValue(item.Name, fieldPathExpression)
+		case item.KubernetesVersion != nil:
+			if item.KubernetesVersion.Version == "" { // I wish there was a ternary operator in Go
```
+1

joe-kimmel-vmw commented on pkg/template/values_factory.go:
```diff
@@ -0,0 +1,22 @@
+	"k8s.io/client-go/kubernetes"
+)
+
+// ValuesFactory abstracts the factories for fetching the values away from the template factory
```
lol wat

neil-hickey replied:
those are definitely words I must have written. They make no sense. LOL

joe-kimmel-vmw reviewed (approved):
I have some comments and questions but i wouldn't argue with merging this as-is.  we can catch up real quick tomorrow AM about whether it's worth addressing any of these in this iteration.

joaopapereira commented on pkg/template/helm_template.go:
```diff
@@ -71,8 +75,15 @@ func (t *HelmTemplate) template(dirPath string, input io.Reader) exec.CmdRunResu
 	args := []string{"template", name, chartPath, "--namespace", namespace, "--include-crds"}
 
+	if t.opts.KubernetesVersion != nil {
+		args = append(args, []string{"--kube-version", t.valuesFactory.additionaDownwardAPIValues.KubernetesVersion.String()}...)
```
```suggestion
		args = append(args, []string{"--kube-version", vals.DownwardAPI().KubernetesVersion}...)
```

My suggestion here is to promote the AdditionalDownwardAPIValues to a primary concern in the Values struct.
This struct should contain all the values that you might eventually need to add to the templating phase.
Also given that the values are supposed to be all strings we should not need to do a `.String()` call when we are using this from the Values struct

neil-hickey replied:
@joaopapereira I moved things back into the Values struct, lemme know if I misunderstood the ask or if it's better now. Honestly, I am fine with it either way, I guess this one is more intentional though

joaopapereira commented on pkg/template/helm_template.go:
```diff
@@ -71,8 +75,15 @@ func (t *HelmTemplate) template(dirPath string, input io.Reader) exec.CmdRunResu
+
 	{ // Add values files
-		vals := Values{t.opts.ValuesFrom, t.appContext, t.coreClient}
+		vals := t.valuesFactory.NewValues(t.opts.ValuesFrom, t.appContext)
```
Move this line to line 77 to accommodate my previous comment

joaopapereira commented on pkg/template/values.go:
```diff
@@ -18,8 +18,9 @@ import (
+	additionalDownwardAPIValues AdditionalDownwardAPIValues
 }
 
 func (t Values) AsPaths(dirPath string) ([]string, func(), error) {
```
Just a comment about this function and maybe not be actionable in this PR.
This signature feels strange, the main reason for that is the return of that `func()` that you need to read the code to understand that it is the removal of the created folders.
It is the responsibility of the creator of `dirPath` to delete that folder, while this function is only responsible for creating the folders it needs inside the `dirPath`.

neil-hickey replied:
Not going to action on this one in this PR

joaopapereira reviewed (commented):
Added some comments

neil-hickey committed:
Address review comments

- renaming some things
- moved away from a values factory back to values struct
- minor fixups

Signed-off-by: Neil Hickey <nhickey@vmware.com>

cppforlife commented on pkg/app/app_template.go:
```diff
@@ -26,20 +26,40 @@ func (a *App) template(dirPath string) exec.CmdRunResult {
 	var result exec.CmdRunResult
 	var isStream bool
 
+	kubernetesVersion, err := a.compInfo.KubernetesVersion(a.app.Spec.ServiceAccountName, a.app.Spec.Cluster, &a.app.ObjectMeta)
+	if err != nil {
+		result.AttachErrorf("%s", fmt.Errorf("Unable to get kubernetes version before templating"))
+		return result
+	}
+
+	kappControllerVersion, err := a.compInfo.KappControllerVersion()
+	if err != nil {
+		result.AttachErrorf("%s", fmt.Errorf("Unable to get kapp-controller version before templating"))
+		return result
+	}
+
+	k8sAPIs, err := a.compInfo.KubernetesAPIs()
+	if err != nil {
+		result.AttachErrorf("%s", fmt.Errorf("Unable to list all server apigroups/version before templating"))
+		return result
+	}
+
```
perf: we should not have to retrieve any of this info if it's not being consumed by values.

neil-hickey replied:
Hmm, that's an interesting ask. So the reason it's like it is currently is:

1. I wanted to call each of these APIs only once before every templating step. Mainly because we have no good way to cache the k8s version on an external cluster, nor the k8s apis. We could do as Joe suggested with a one minute cache. 
2. The calls would have to get moved down into each factory individually (i.e duplicate the information gathering of something that ought not to change between templating steps.

neil-hickey replied:
I went ahead and memoized the calls, I added a test which validates the calls are made made only once with multiple template steps. pkg/app/app_template_test.go

cppforlife commented on pkg/app/app_template.go:
```diff
@@ -26,20 +26,40 @@ func (a *App) template(dirPath string) exec.CmdRunResult {
 
+	kubernetesVersion, err := a.compInfo.KubernetesVersion(a.app.Spec.ServiceAccountName, a.app.Spec.Cluster, &a.app.ObjectMeta)
+	if err != nil {
+		result.AttachErrorf("%s", fmt.Errorf("Unable to get kubernetes version before templating"))
```
should be including underlying err

cppforlife commented on pkg/packageinstall/packageinstall.go:
```diff
@@ -248,12 +241,14 @@ func (pi *PackageInstallCR) kcVersionConstraintsSatisfied(pkg *datapkgingv1alpha
 		return false
 	}
 
+	v.Pre = semver.PRVersion{}
+	v.Build = semver.BuildMeta{}
```
should this move into ComponentInfo? seems its more related to version parsing rather than package matching

neil-hickey replied:
The constraint matching is specific to packageinstall (i.e the choice to drop the pre and build fields of KC version), so I think it makes more sense here. The kapp-controller version should always be as it is in the cluster, regardless of how / why we manipulate it in kapp-controller.

joe-kimmel-vmw replied:
I understand the point that it's not entirely semantically in-scope for the reader of PackageInstall to know about the effects of resetting the members of the version, `v` struct.   However it's less obvious to me where to move that logic, and i think it'll be a similar number of lines of code,  just maybe a little faster to read packageinstall.go until your concern is this logic specifically.     To the extent the concern is just readability, it feels like a refactor or an explanatory comment would have similar benefits.

cppforlife commented on pkg/apis/kappctrl/v1alpha1/types_template.go:
```diff
@@ -53,6 +53,11 @@ type AppTemplateHelmTemplate struct {
+	// Can be manually overridden to a value instead.
+	KubernetesVersion *Version `json:"kubernetesVersion,omitempty" protobuf:"bytes,5,opt,name=kubernetesVersion"`
+	// Optional: Use kubernetes group/versions resources available in the live cluster
+	KubernetesAPIs *Version `json:"kubernetesAPIs,omitempty" protobuf:"bytes,6,opt,name=kubernetesAPIs"`
```
seems like our test isnt catching that this shouldnt be a *Version.

cppforlife commented on pkg/apis/kappctrl/v1alpha1/types_template.go:
```diff
@@ -68,16 +73,35 @@ type AppTemplateValuesSourceRef struct {
 // +k8s:openapi-gen=true
-type AppTemplateValuesDownwardAPI struct {
-	Items []AppTemplateValuesDownwardAPIItem `json:"items,omitempty" protobuf:"bytes,1,opt,name=items"`
+type Version struct {
```
i would recommend renaming this to KubernetesVersion and KubernetesGroupVersions

neil-hickey replied:
Why exactly? I think `Version` can be used for both KC and K8s version info. If we make it explicitly `KubernetesVersion` wouldn't it look like this?

```
kubernetesVersion:
  kubernetesVersion: "version"

kappControllerVersion:
  kappControllerVersion: "version"
```

Otherwise we have two structs `KappControllerVersion` and `KubernetesVersion` which both match on the json `version` which feels a little redundant, i.e.

```
kubernetesVersion:
  version: "version"

kappControllerVersion:
  version: "version"
```

neil-hickey replied:
I renamed GroupVersions to KubernetesGroupVersions but I left the other as Version as it made more sense to me to keep them together / and it's a key underneath `KubernetesVersion` so it felt fine to leave as is

neil-hickey committed:
Fixup errors in app_template

Signed-off-by: Neil Hickey <nhickey@vmware.com>

neil-hickey committed:
Add memoized fetching of versions

- Add template() test to validate memoizing works

Signed-off-by: Neil Hickey <nhickey@vmware.com>

joe-kimmel-vmw commented on pkg/packageinstall/packageinstall.go:
```diff
@@ -301,7 +301,18 @@ func (pi *PackageInstallCR) referencedPkgVersion() (datapkgingv1alpha1.Package,
 	if requiresClusterVersion {
-		v, err := pi.compInfo.KubernetesVersion(pi.model.Spec.ServiceAccountName, pi.model.Spec.Cluster, &pi.model.ObjectMeta)
+		kubernetesVersion := func() (semver.Version, error) {
+			memoizedKubernetesVersion := semver.Version{}
```
here the naming suggests you're memoizing but the variable doesn't have scope outside the function, right?  am I missing how this would be persisted across function calls?   (vs. a couple files up where you declared the variables in the outer scope and then assigned to them in the func)

Commit:
remove unnecessary memoization

- memoizing within packageinstall didnt actually memoize
- memoizing within componentinfo was too aggressive so controller would not receive updated version after cluster is updated

Commit:
rename kubernetesGroupVersions to kubernetesAPIs

Commit:
use array as type of values for kubernetesAPIs
